app/src/main/java/org/openobservatory/ooniprobe/client/callback/GetMeasurementsCallback.java

```diff
@@ -0,0 +1,31 @@
+package org.openobservatory.ooniprobe.client.callback;
+
+import org.openobservatory.ooniprobe.model.api.ApiMeasurement;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public abstract class GetMeasurementsCallback implements Callback<ApiMeasurement> {
+    @Override
+    public void onResponse(Call<ApiMeasurement> call, Response<ApiMeasurement> response) {
+        if (response.isSuccessful() &&
+                response.body() != null &&
+                response.body().results != null &&
+                response.body().results.size() == 1 &&
+                response.body().results.get(0).measurement_url != null) {
+            onSuccess(response.body().results.get(0));
+        } else {
+            onError(Integer.toString(response.code()));
```

Comment:
I believe this `onError` clause may trigger for `200` status code if the response JSON does not contain all the information that we expect it to have. I suggest to change the error to mention that it's either an HTTP error or a malformed JSON. (We can still include the response code into the JSON.) &mdash; Or am I missing something?

Reply:
For example, I see you're using a converter factory. I wonder what the error code is if the JSON does not contain all the field that we expect. And what if it is malformed instead?

Reply:
All the checks are done, array exist, array is exactly one in size and the object 0 has the field we need, exactly like iOS.
```
        if (response.isSuccessful() &&
                response.body() != null &&
                response.body().results != null &&
                response.body().results.size() == 1 &&
                response.body().results.get(0).measurement_url != null) 
```

Reply:
Right. The question is: suppose response is okay but somehow `body` is `null`. Then you get an error with, I assume, error code `200`. Is that helpful?

Reply:
At the moment (in iOS) we always show a custom string message, saying that we couldn't retrieve the meassurement with a custom error code.

Reply:
Right; so, can we have something more meaningful also here? (I'm trying to advocate for avoiding errors that also developers would find obscure.

Reply:
If you want to suggest a different approach just say what you want to achieve and we'll have to change the iOS build as well

Reply:
Wait, do we return, say `400` as error on iOS?

Reply:
We return this
`#define ERR_JSON_EMPTY 100`

Reply:
We are not required to force returning the same type on iOS and Android.

What is the semantic of the error being returned here? How do you indicate what went wrong? What is the point of returning the status code, which is an unrelated bit of information, as the error code?

On a more general note, what is our objective here? Perhaps I am not understanding the objective

Reply:
on iOS we return a custom error code (100) with an error (json empty)
Here we return the returned error code from the web with the same message (json empty)

Reply:
We do:

```Java
    onError(Integer.toString(response.code()));
```

Therefore:

> the returned error code from the web

This is part of what I find confusing. There may be _no_ error code from the web. The web API may return `200` (meaning OK) with a body that for some reasons does not match the expectations.

> with the same message (json empty)

I do not see any message here. You just stringify the integer error code.

To cut down the length of the discussion, I'll propose two way forward for this:

# 1. Precise error handling

```Java
if (!response.isSuccessful()) {
  onError("The web API failed with status code: " + Integer.toString(response.code()));
  return;
}
if (response.body() == null) {
  onError("We have a null response body");
  return;
}
if (response.body().results == null) {
  onError("We have no results field in the body");
  return;
}
if (response.body().results.size() != 1) {
  onError("We have too many results");
  return;
}
if (response.body().results.get(0).measurement_url == null) {
  onError("There is no measurement url");
  return;
}
onSuccess(response.body().results.get(0));
```

# 2. Predictable, coarse grained error handling

```Java
if (response.isSuccessful() &&
    response.body() != null &&
    response.body().results != null &&
    response.body().results.size() == 1 &&
    response.body().results.get(0).measurement_url != null) {
      onSuccess(response.body().results.get(0));
} else {
    onError("Request failed or invalid body; HTTP code was: " + Integer.toString(response.code()));
}
```

- - -

I would suggest to pick one of these to solve the issues that I see (mainly: the reported error is not clear).

Reply:
At the moment I will return `R.string.Modal_Error_JsonEmpty` as we do in iOS

Reply:
Sounds good to me!